model/DataObject.php

```diff
@@ -1876,164 +1932,256 @@ public function has_many($component = null, $classOnly = true) {
 	}
 
 	/**
+	 * Return data for a specific has_many component.
+	 * @param string $component
+	 * @param bool $classOnly If this is TRUE, than any has_many relationships in the form "ClassName.Field" will have
+	 *        the field data stripped off. It defaults to TRUE.
+	 * @return string|false
+	 */
+	public function hasManyComponent($component, $classOnly = true) {
+		$hasMany = $this->stat('has_many');
+		if($hasMany && array_key_exists($component, $hasMany)) {
+			$hasMany = $hasMany[$component];
+		} else {
+			return false;
+		}
+
+		return ($classOnly) ? preg_replace('/(.+)?\..+/', '$1', $hasMany) : $hasMany;
+	}
+
+	/**
+	 * @deprecated 4.0 Method has been renamed to manyManyExtraFields()
+	 * @param string $component
+	 * @return array
+	 */
+	public function many_many_extraFields($component = null) {
+		Deprecation::notice('4.0', 'Method has been renamed to manyManyExtraFields()');
+		return $this->manyManyExtraFields($component);
+	}
+
+	/**
 	 * Return the many-to-many extra fields specification.
 	 *
 	 * If you don't specify a component name, it returns all
 	 * extra fields for all components available.
 	 *
 	 * @param string $component Name of component
-	 * @return array
+	 * @return array|null
 	 */
-	public function many_many_extraFields($component = null) {
-		$classes = ClassInfo::ancestry($this);
+	public function manyManyExtraFields($component = null) {
+		if($component) {
+			return $this->manyManyExtraFieldsForComponent($component);
+		}
 
+		$classes = ClassInfo::ancestry($this);
 		foreach($classes as $class) {
 			if(in_array($class, array('ViewableData', 'Object', 'DataObject'))) continue;
-			$relationName = null;
 
-			// Find extra fields for one component
-			if($component) {
-				$SNG_class = singleton($class);
-				$extraFields = $SNG_class->stat('many_many_extraFields');
+			// Find all the extra fields for all components
+			$newItems = eval("return (array){$class}::\$many_many_extraFields;");
```

Comment:
I think given all this work the merging will be a nightmare and the other fix is likely to NOT be merged in if you don't do it now.